Source/Core/VideoCommon/TextureCacheBase.cpp

```diff
@@ -293,6 +295,318 @@ void TextureCacheBase::ScaleTextureCacheEntryTo(TextureCacheBase::TCacheEntry* e
+    DoSaveState(p);
+  else
+    DoLoadState(p);
+}
+
+void TextureCacheBase::DoSaveState(PointerWrap& p)
+{
+  std::map<TCacheEntry*, u32> entry_map;
+  auto AddCacheEntryToMap = [&entry_map, &p](TCacheEntry* entry) {
+    auto iter = entry_map.find(entry);
+    if (iter != entry_map.end())
+      return;
+
+    // Skip custom textures since they won't fit in our staging texture.
+    if (entry->is_custom_tex)
+      return;
+
+    u32 id = static_cast<u32>(entry_map.size());
+    entry_map.emplace(entry, id);
+  };
+  auto GetCacheEntryId = [&entry_map](TCacheEntry* entry) -> u32 {
+    auto iter = entry_map.find(entry);
+    return iter != entry_map.end() ? iter->second : static_cast<u32>(-1);
+  };
+
+  // Collect all referenced texture cache entries and assign them IDs.
+  for (const auto& it : textures_by_address)
+    AddCacheEntryToMap(it.second);
+  for (const auto& it : textures_by_hash)
+    AddCacheEntryToMap(it.second);
+
+  // Save the texture cache entries out in the order the were referenced.
+  u32 size = static_cast<u32>(entry_map.size());
+  p.Do(size);
+  for (const auto& it : entry_map)
+  {
+    TCacheEntry* entry = it.first;
+    TCacheEntry::DoState(p, &entry);
+  }
+  p.DoMarker("TextureCacheEntries");
+
+  // Save references for each cache entry.
+  // As references are circular, we need to have everything created before linking entries.
+  std::set<std::pair<u32, u32>> reference_pairs;
+  for (const auto& it : entry_map)
+  {
+    TCacheEntry* entry = it.first;
+    u32 id1 = GetCacheEntryId(entry);
+    for (TCacheEntry* referenced_entry : entry->references)
+    {
+      u32 id2 = GetCacheEntryId(referenced_entry);
+      auto refpair1 = std::make_pair(id1, id2);
+      auto refpair2 = std::make_pair(id2, id1);
+      if (reference_pairs.count(refpair1) == 0 && reference_pairs.count(refpair2) == 0)
+        reference_pairs.insert(refpair1);
+    }
+  }
+  size = static_cast<u32>(reference_pairs.size());
+  p.Do(size);
+  for (const auto& it : reference_pairs)
+  {
+    p.Do(it.first);
+    p.Do(it.second);
+  }
+  p.DoMarker("TextureCacheReferencePairs");
+
+  // Save the address maps.
+  size = static_cast<u32>(textures_by_address.size());
+  p.Do(size);
+  for (auto& it : textures_by_address)
+  {
+    u32 id = GetCacheEntryId(it.second);
+    p.Do(it.first);
+    p.Do(id);
+  }
+  size = static_cast<u32>(textures_by_hash.size());
+  p.Do(size);
+  for (auto& it : textures_by_hash)
+  {
+    u32 id = GetCacheEntryId(it.second);
+    p.Do(it.first);
+    p.Do(id);
+  }
+
+  // Free the readback texture to potentially save host-mapped GPU memory, depending on where
+  // the driver mapped the staging buffer.
+  m_readback_texture.reset();
+}
+
+void TextureCacheBase::DoLoadState(PointerWrap& p)
+{
+  // Helper for getting a cache entry from an ID.
+  std::map<u32, TCacheEntry*> id_map;
+  auto GetEntry = [&id_map](u32 id) {
+    auto iter = id_map.find(id);
+    return iter == id_map.end() ? nullptr : iter->second;
```

Comment:
Wouldn't returning an optional make more sense here to signify lack of a mapping?